# Get data

In [1]:
# imports
import pandas as pd

In [2]:
monthly = pd.read_excel('Sales by part by month.xlsx', sheet_name='Sheet3')
monthly.head(20)

Part  Unit Price  Quantity  Customer     InvNo  Final Price    EntDate  \
0    8812-2         0.0         4    409060  14800000          0.0 2012-04-03   
1    7267-2        30.0        -1    404230  14800001        -30.0 2012-04-04   
2    2922-7         0.0         8    408130  14800082          0.0 2012-04-06   
3    2944-7        12.0         6    403980  14800071         72.0 2012-04-06   
4    2944-7        15.0         4    403710  14800072         60.0 2012-04-06   
5      4472        40.0         2    400590  14800051         80.0 2012-04-06   
6    4472-7        40.0         1    402390  14800093         40.0 2012-04-06   
7      4476        40.0         1    400590  14800051         40.0 2012-04-06   
8      4483        31.9         2    403830  14800085         63.8 2012-04-06   
9      8861        13.0         3    407510  14800053         39.0 2012-04-06   
10   7250-1         0.0         1    413100  14800015          0.0 2012-04-06   
11   7267-1        20.0         1    403710  14800072         20.0 2012-04-06   
12   7267-3         0.0         1    406900  14800002          0.0 2012-04-06   
13     8000        10.5         1    404150  14800055         10.5 2012-04-06   
14     8000        10.5         5    407510  14800053         52.5 2012-04-06   
15  7372/12         0.0         8    408130  14800082          0.0 2012-04-06   
16   7372/8        16.0         1    400590  14800051         16.0 2012-04-06   
17   7372/8         0.0         2    413040  14800058          0.0 2012-04-06   
18  7379/10        36.0         6    404900  14800057        216.0 2012-04-06   
19   5558-1        10.0         6    407130  14800052         60.0 2012-04-06   

      PLn  Agt    InvDate      RecordID  InvDate.1 InvDate 2 InvDate3  
0   50612   70 2012-04-03  R|51236|0|10 2012-04-03  Apr-2012  2012-04  
1   20202   61 2012-04-04  R|51237|0|10 2012-04-04  Apr-2012  2012-04  
2   20008   99 2012-04-06  R|51319|0|10 2012-04-06  Apr-2012  2012-04  
3   20008   15 2012-04-06  R|51308|0|10 2012-04-06  Apr-2012  2012-04  
4   20008   15 2012-04-06  R|51309|0|10 2012-04-06  Apr-2012  2012-04  
5   20008   20 2012-04-06  R|51288|0|50 2012-04-06  Apr-2012  2012-04  
6   20008   65 2012-04-06  R|51330|0|90 2012-04-06  Apr-2012  2012-04  
7   20008   22 2012-04-06  R|51288|0|60 2012-04-06  Apr-2012  2012-04  
8   20008    6 2012-04-06  R|51322|0|20 2012-04-06  Apr-2012  2012-04  
9   20008   16 2012-04-06  R|51290|0|90 2012-04-06  Apr-2012  2012-04  
10  20202   61 2012-04-06  R|51252|0|10 2012-04-06  Apr-2012  2012-04  
11  20202   99 2012-04-06  R|51309|0|30 2012-04-06  Apr-2012  2012-04  
12  20202    6 2012-04-06  R|51238|0|50 2012-04-06  Apr-2012  2012-04  
13  20202   22 2012-04-06  R|51292|0|30 2012-04-06  Apr-2012  2012-04  
14  20202   22 2012-04-06  R|51290|0|80 2012-04-06  Apr-2012  2012-04  
15  20203   65 2012-04-06  R|51319|0|80 2012-04-06  Apr-2012  2012-04  
16  20203   70 2012-04-06  R|51288|0|80 2012-04-06  Apr-2012  2012-04  
17  20203   70 2012-04-06  R|51295|0|20 2012-04-06  Apr-2012  2012-04  
18  20203   20 2012-04-06  R|51294|0|10 2012-04-06  Apr-2012  2012-04  
19  20301   99 2012-04-06  R|51289|0|30 2012-04-06  Apr-2012  2012-04

In [3]:
monthly.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 922077 entries, 0 to 922076
Data columns (total 14 columns):
Part           922077 non-null object
Unit Price     922077 non-null float64
Quantity       922077 non-null int64
Customer       922077 non-null int64
InvNo          922077 non-null int64
Final Price    922077 non-null float64
EntDate        922077 non-null datetime64[ns]
PLn            922077 non-null int64
Agt            922077 non-null int64
InvDate        922077 non-null datetime64[ns]
RecordID       922077 non-null object
InvDate.1      922077 non-null datetime64[ns]
InvDate 2      922077 non-null object
InvDate3       922077 non-null object
dtypes: datetime64[ns](3), float64(2), int64(5), object(4)
memory usage: 98.5+ MB


In [4]:
monthly.describe()

Unit Price       Quantity       Customer         InvNo  \
count  922077.00000  922077.000000  922077.000000  9.220770e+05   
mean       48.27954      10.679287  405662.277386  1.438898e+07   
std       239.54981      59.570260    4080.905777  3.886682e+05   
min         0.00000   -1030.000000  400000.000000  1.400000e+07   
25%        11.00000       1.000000  402120.000000  1.406094e+07   
50%        30.00000       2.000000  404820.000000  1.412024e+07   
75%        55.00000       6.000000  407910.000000  1.484688e+07   
max    218185.40000   10000.000000  499999.000000  1.491743e+07   

         Final Price            PLn            Agt  
count  922077.000000  922077.000000  922077.000000  
mean      397.883230   45225.684241      40.027002  
std      2420.515708   11462.769806      32.747810  
min   -199699.500000   20008.000000       5.000000  
25%        30.000000   50101.000000      15.000000  
50%        80.000000   50112.000000      20.000000  
75%       184.680000   50707.000000      67.000000  
max    240000.000000   52112.000000      99.000000

QUESTIONS to Lee:
* What do minus quantities mean? Voided order?
* How are EntDate and InvDate different? Do you prefer one over the other?
* Items with dashes (e.g. 4472 and 4472-7) are different items?
* There are a few items that seem fat-fingered entries. Maximum unit price is 218,185??? and quantities interquartile range is 1-6 but minimum is -1030 and maximum is 10,000?
* What is PLn or Agt? (are they meaningful?)

good news: no null values! no categorical data or text! yay!!!

In [5]:
# rename Unit Price
monthly['Unit_Price']=monthly['Unit Price']
monthly.head()

Part  Unit Price  Quantity  Customer     InvNo  Final Price    EntDate  \
0  8812-2         0.0         4    409060  14800000          0.0 2012-04-03   
1  7267-2        30.0        -1    404230  14800001        -30.0 2012-04-04   
2  2922-7         0.0         8    408130  14800082          0.0 2012-04-06   
3  2944-7        12.0         6    403980  14800071         72.0 2012-04-06   
4  2944-7        15.0         4    403710  14800072         60.0 2012-04-06   

     PLn  Agt    InvDate      RecordID  InvDate.1 InvDate 2 InvDate3  \
0  50612   70 2012-04-03  R|51236|0|10 2012-04-03  Apr-2012  2012-04   
1  20202   61 2012-04-04  R|51237|0|10 2012-04-04  Apr-2012  2012-04   
2  20008   99 2012-04-06  R|51319|0|10 2012-04-06  Apr-2012  2012-04   
3  20008   15 2012-04-06  R|51308|0|10 2012-04-06  Apr-2012  2012-04   
4  20008   15 2012-04-06  R|51309|0|10 2012-04-06  Apr-2012  2012-04   

   Unit_Price  
0         0.0  
1        30.0  
2         0.0  
3        12.0  
4        15.0

In [6]:
# create a dataframe where EntDate or InvDate is the index
# columns are: Part, Customer, Unit Price - will add promoted yes/no as a binary later
cols = ['EntDate', 'Part','Customer', 'Unit_Price', 'Quantity']
data = monthly[cols]
data.head(20)

EntDate     Part  Customer  Unit_Price  Quantity
0  2012-04-03   8812-2    409060         0.0         4
1  2012-04-04   7267-2    404230        30.0        -1
2  2012-04-06   2922-7    408130         0.0         8
3  2012-04-06   2944-7    403980        12.0         6
4  2012-04-06   2944-7    403710        15.0         4
5  2012-04-06     4472    400590        40.0         2
6  2012-04-06   4472-7    402390        40.0         1
7  2012-04-06     4476    400590        40.0         1
8  2012-04-06     4483    403830        31.9         2
9  2012-04-06     8861    407510        13.0         3
10 2012-04-06   7250-1    413100         0.0         1
11 2012-04-06   7267-1    403710        20.0         1
12 2012-04-06   7267-3    406900         0.0         1
13 2012-04-06     8000    404150        10.5         1
14 2012-04-06     8000    407510        10.5         5
15 2012-04-06  7372/12    408130         0.0         8
16 2012-04-06   7372/8    400590        16.0         1
17 2012-04-06   7372/8    413040         0.0         2
18 2012-04-06  7379/10    404900        36.0         6
19 2012-04-06   5558-1    407130        10.0         6

In [7]:
data.set_index('EntDate', inplace=True)
by_week = data.groupby(['Part']).resample('W')
by_week.Quantity.sum().head(20)

Part  EntDate   
012   2013-01-27     1
017   2012-08-19     1
      2012-08-26     0
      2012-09-02     0
      2012-09-09     0
      2012-09-16     0
      2012-09-23     0
      2012-09-30     0
      2012-10-07     0
      2012-10-14     0
      2012-10-21     0
      2012-10-28     0
      2012-11-04    48
      2012-11-11     0
      2012-11-18    18
      2012-11-25     0
      2012-12-02     0
      2012-12-09     0
      2012-12-16     0
      2012-12-23     0
Name: Quantity, dtype: int64

In [13]:
df = by_week.Quantity.sum().unstack(1).fillna(0)

In [18]:
df.head(20)

EntDate       2012-04-08  2012-04-15  2012-04-22  2012-04-29  2012-05-06  \
Part                                                                       
012                  0.0         0.0         0.0         0.0         0.0   
017                  0.0         0.0         0.0         0.0         0.0   
03008944ST-1         0.0         0.0         0.0         0.0         0.0   
03008944ST-3         0.0         0.0         0.0         0.0         0.0   
0300ST1550-1         0.0         0.0         0.0         0.0         0.0   
0300ST15X9-1         0.0         0.0         0.0         0.0         0.0   
0300ST15X9-2         0.0         0.0         0.0         0.0         0.0   
0300ST15X9-3         0.0         0.0         0.0         0.0         0.0   
0300ST1605-1         0.0         0.0         0.0         0.0         0.0   
0300ST1605-2         0.0         0.0         0.0         0.0         0.0   
0300ST245-1          0.0         0.0         0.0         0.0         0.0   
0300ST245-2          0.0         0.0         0.0         0.0         0.0   
0300ST320-1          0.0         0.0         0.0         0.0         0.0   
0300ST320-2          0.0         0.0         0.0         0.0         0.0   
0300ST320-3          0.0         0.0         0.0         0.0         0.0   
0300ST365-1          0.0         1.0         0.0       200.0         0.0   
0300ST365-2          0.0         0.0         0.0         0.0         0.0   
0300ST365-3          0.0         1.0         0.0         0.0         0.0   
0300ST450-1          0.0         2.0         0.0         0.0         0.0   
0300ST450-2          0.0         0.0         0.0         0.0         0.0   

EntDate       2012-05-13  2012-05-20  2012-05-27  2012-06-03  2012-06-10  \
Part                                                                       
012                  0.0         0.0         0.0         0.0         0.0   
017                  0.0         0.0         0.0         0.0         0.0   
03008944ST-1         0.0         0.0         0.0         0.0         0.0   
03008944ST-3         0.0         0.0         0.0         0.0         0.0   
0300ST1550-1         0.0         0.0         0.0         0.0         0.0   
0300ST15X9-1         0.0         0.0         0.0         0.0         0.0   
0300ST15X9-2         0.0         0.0         0.0         0.0         0.0   
0300ST15X9-3         0.0         0.0         0.0         0.0         0.0   
0300ST1605-1         0.0         0.0         0.0         0.0         0.0   
0300ST1605-2         0.0         0.0         0.0         0.0         0.0   
0300ST245-1          0.0         0.0         0.0         0.0         0.0   
0300ST245-2          0.0         0.0         0.0         0.0         0.0   
0300ST320-1          0.0         0.0         0.0         0.0         0.0   
0300ST320-2          0.0         0.0         0.0         0.0         0.0   
0300ST320-3          0.0         0.0         0.0         0.0         0.0   
0300ST365-1          0.0         5.0         0.0         0.0         0.0   
0300ST365-2          0.0         0.0         0.0         0.0         0.0   
0300ST365-3          0.0         0.0         0.0         0.0         0.0   
0300ST450-1          0.0         0.0         0.0         0.0         0.0   
0300ST450-2          0.0         0.0         0.0         0.0         0.0   

EntDate          ...      2017-09-03  2017-09-10  2017-09-17  2017-09-24  \
Part             ...                                                       
012              ...             0.0         0.0         0.0         0.0   
017              ...             0.0         0.0         0.0         0.0   
03008944ST-1     ...             0.0         0.0         0.0         0.0   
03008944ST-3     ...             0.0         0.0         0.0         0.0   
0300ST1550-1     ...             0.0         0.0         0.0         0.0   
0300ST15X9-1     ...             0.0         0.0         0.0         0.0   
0300ST15X9-2     ...             0.0         0

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1833 entries, 012 to TRUHONE
Columns: 292 entries, 2012-04-08 to 2017-11-05
dtypes: float64(292)
memory usage: 4.1+ MB


In [19]:
df_transposed = df.T
df_transposed.head(20)

Part        012  017  03008944ST-1  03008944ST-3  0300ST1550-1  0300ST15X9-1  \
EntDate                                                                        
2012-04-08  0.0  0.0           0.0           0.0           0.0           0.0   
2012-04-15  0.0  0.0           0.0           0.0           0.0           0.0   
2012-04-22  0.0  0.0           0.0           0.0           0.0           0.0   
2012-04-29  0.0  0.0           0.0           0.0           0.0           0.0   
2012-05-06  0.0  0.0           0.0           0.0           0.0           0.0   
2012-05-13  0.0  0.0           0.0           0.0           0.0           0.0   
2012-05-20  0.0  0.0           0.0           0.0           0.0           0.0   
2012-05-27  0.0  0.0           0.0           0.0           0.0           0.0   
2012-06-03  0.0  0.0           0.0           0.0           0.0           0.0   
2012-06-10  0.0  0.0           0.0           0.0           0.0           0.0   
2012-06-17  0.0  0.0           0.0           0.0           0.0           0.0   
2012-06-24  0.0  0.0           0.0           0.0           0.0           0.0   
2012-07-01  0.0  0.0           0.0           0.0           0.0           0.0   
2012-07-08  0.0  0.0           0.0           0.0           0.0           0.0   
2012-07-15  0.0  0.0           0.0           0.0           0.0           0.0   
2012-07-22  0.0  0.0           0.0           0.0           0.0           0.0   
2012-07-29  0.0  0.0           0.0           0.0           0.0           0.0   
2012-08-05  0.0  0.0           0.0           0.0           0.0           0.0   
2012-08-12  0.0  0.0           0.0           0.0           0.0           0.0   
2012-08-19  0.0  1.0           0.0           0.0           0.0           0.0   

Part        0300ST15X9-2  0300ST15X9-3  0300ST1605-1  0300ST1605-2   ...     \
EntDate                                                              ...      
2012-04-08           0.0           0.0           0.0           0.0   ...      
2012-04-15           0.0           0.0           0.0           0.0   ...      
2012-04-22           0.0           0.0           0.0           0.0   ...      
2012-04-29           0.0           0.0           0.0           0.0   ...      
2012-05-06           0.0           0.0           0.0           0.0   ...      
2012-05-13           0.0           0.0           0.0           0.0   ...      
2012-05-20           0.0           0.0           0.0           0.0   ...      
2012-05-27           0.0           0.0           0.0           0.0   ...      
2012-06-03           0.0           0.0           0.0           0.0   ...      
2012-06-10           0.0           0.0           0.0           0.0   ...      
2012-06-17           0.0           0.0           0.0           0.0   ...      
2012-06-24           0.0           0.0           0.0           0.0   ...      
2012-07-01           0.0           0.0           0.0           0.0   ...      
2012-07-08           0.0           0.0           0.0           0.0   ...      
2012-07-15           0.0           0.0           0.0           0.0   ...      
2012-07-22           0.0           0.0           0.0           0.0   ...      
2012-07-29           0.0           0.0           0.0           0.0   ...      
2012-08-05           0.0           0.0           0.0           0.0   ...      
2012-08-12           0.0           0.0           0.0           0.0   ...      
2012-08-19           0.0           0.0           0.0           0.0   ...      

Part        9920-2  9920-3  9920-4  9920-5  9920-6  9920-7  9997-25  HW220D15  \
EntDate                                                                         
2012-04-08     0.0     0.0     0.0     0.0     0.0     0.0      0.0       0.0   
2012-04-15    20.0    19.0     7.0   166.0    98.0     0.0      0.0       0.0   
2012-04-22    41.0    15.0    10.0   207.0    87.0     0.0      0.0       0.0   
2012-04-29    38.0    44.0    13.0   101.0    21.0    10.0      0.0       0.0   
2012-05-06    

In [20]:
df_transposed.to_excel('wusthof_time_series.xlsx')